fgpyo/io/__init__.py

```diff
@@ -0,0 +1,200 @@
+"""
+
+import gzip
+import io
+import os
+from pathlib import Path
+from typing import IO
+from typing import Any
+from typing import Iterable
+from typing import Iterator
+from typing import Set
+from typing import TextIO
+from typing import Union
+from typing import cast
+
+COMPRESSED_FILE_EXTENSIONS: Set[str] = {".gz", ".bgz"}
+
+
+def assert_path_is_readable(path: Path) -> None:
+    """Checks that file exists and returns True, else raises AssertionError
+
+    Args:
+        path: a Path to check
+
+    Example:
+    >>> assert_file_exists(path = Path("some_file.csv"))
+    """
+    assert path.exists(), f"Cannot read non-existent path: {path}"
+    assert not path.is_dir(), f"Cannot read path becasue it is a directory: {path}"
+    assert os.access(path, os.R_OK), f"Path exists but is not readable: {path}"
+
+
+def assert_directory_exists(path: Path) -> None:
+    """Asserts that a path exist and is a directory
+
+    Args:
+        path: Path to check
+
+    Example:
+    >>> assert_directory_exists(path = Path("/example/directory/"))
+    """
+    assert path.exists(), f"Path does not exist: {path}"
+    assert path.is_dir(), f"Path exists but is not a directory: {path}"
+
+
+def assert_path_is_writeable(path: Path, parent_must_exist: bool = True) -> None:
+    """Asserts the following:
+    If the file exists then it must also be writeable. Else if the path is not a file and
+    parent_must_exist is true then assert that the parent directory exists and is writeable.
+    Else if the path is not a directory and parent_must_exist is false then look at each parent
+    directory until one is found that exists and is writeable.
+
+    Args:
+         path: Path to check
+         parent_must_exist: True/False the parent directory must exist, the default is True
+
+    Example:
+    >>> assert_path_is_writeable(path = Path("example.txt"))
+    """
+    # If file exists, it must be writeable
+    if path.exists():
+        assert path.is_file() and os.access(
+            path, os.W_OK
+        ), f"File exists but is not writebale: {path}"
+        # return None
+
+    # Else if file doesnt exist and parent_must_exist is True then check
+    # that path.absolute().parent exists, is a directory and is writeable
+    elif parent_must_exist:
+        assert (
+            path.absolute().parent.exists()
+            & path.absolute().parent.is_dir()
+            & os.access(path.absolute().parent, os.W_OK)
+        ), f"Path does not exist and parent isn't extent/writable: {path}"
+        # return None
+
+    # Else if file doesn't exist and parent_must_exist is False, test parent until
+    # you find the first extent path, and check that it is a directory and is writeable.
+    else:
+        for parent in path.parents:
+            if parent.exists():
+                assert parent.exists() & os.access(
+                    parent, os.W_OK
+                ), f"File does not have a writeable parent directory: {path}"
+            return
```

Comment:
The return needs to be indented one more level, so it's only triggered under `if parent.exists()` after the assert.